Add CDC partitioner (Version 4.9.0)

avelanarius opened the pull request:
Add new class `CDCToken` which implements a Scylla CDC partitioner, as defined here:

https://github.com/scylladb/scylla/blob/master/cdc/cdc_partitioner.cc

Add new `getPartitioner()` method in `Request` and `PreparedStatement` in order to be able to use custom partitioner on CDC log tables (and potentially other in the future).

Modify `Conversions.java` to set CDCPartitioner for CDC log prepared statements.

Change `TokenMap`, `CqlRequestHandler` to be able to use custom partitioner (or cluster-wide if null is provided).

A new `TokenLong64` type is introduced to allow comparisons between `Murmur3Token` and `CDCToken` (both backed by 64-bit integers), as an existing `TokenMap`(with `Murmur3Token`) is used.

Similarly to https://github.com/scylladb/java-driver/pull/35, I have performed manual testing (running queries with tracing and checking if they are sent to proper node and shard).

avelanarius commented:
@haaawk ping

haaawk commented on core/src/main/java/com/datastax/oss/driver/internal/core/metadata/token/CDCToken.java:
```diff
@@ -0,0 +1,17 @@
+package com.datastax.oss.driver.internal.core.metadata.token;
+
+import net.jcip.annotations.Immutable;
```
I'm not a great fan of nonstandard annotations like this. Could we please remove it?

haaawk commented on core/src/main/java/com/datastax/oss/driver/internal/core/metadata/token/CDCTokenFactory.java:
```diff
@@ -0,0 +1,84 @@
+      return MIN_TOKEN;
+    }
+
+    long upperDword = partitionKey.getLong(offset + 0);
```
offset + 0?

avelanarius replied:
To indicate that `offset` is a starting offset and we will read first long (`upperDword`) and second long (`lowerDword`), explicitly stating the "internal" offsets of those two values.

haaawk replied:
I don't see much value in it but ok.

haaawk commented on core/src/main/java/com/datastax/oss/driver/internal/core/metadata/token/CDCTokenFactory.java:
```diff
@@ -0,0 +1,84 @@
+@ThreadSafe
+public class CDCTokenFactory implements TokenFactory {
+
+  public static final String PARTITIONER_NAME = "com.scylladb.dht.CDCPartitioner";
+
+  public static final CDCToken MIN_TOKEN = new CDCToken(Long.MIN_VALUE);
+  public static final CDCToken MAX_TOKEN = new CDCToken(Long.MAX_VALUE);
+
```
Why are those fields public?

avelanarius replied:
No good reason aside from the fact that other (already implemented) TokenFactories also have them public:

https://github.com/scylladb/java-driver/blob/017b27315aabbc813bfda4070f778cbdb37907ec/core/src/main/java/com/datastax/oss/driver/internal/core/metadata/token/Murmur3TokenFactory.java#L29-L30

(consistency)

haaawk replied:
ok.

haaawk commented on core/src/main/java/com/datastax/oss/driver/internal/core/metadata/token/ByteOrderedTokenFactory.java:
```diff
@@ -16,6 +16,7 @@
 package com.datastax.oss.driver.internal.core.metadata.token;
 
 import com.datastax.oss.driver.api.core.metadata.token.Token;
+import com.datastax.oss.driver.api.core.metadata.token.TokenFactory;
```
Why is this needed when you haven't changed this class?

avelanarius replied:
The whole story:

1. I added `TokenFactory getPartitioner()` to `com.datastax.oss.driver.api.core.cql.PreparedStatement` (`api` NOT `internal` package).
2. After that, automatic checker in build system complained that `TokenFactory` is from `com.datastax.oss.driver.internal.core.metadata.token` (`internal`!). It complained that `api` cannot import stuff from `internal` (probably for automatic docs purposes). Correction: Probably not import but expose "to the external world" - in this case by making `internal` class as a return type.
3. That forced me to move `TokenFactory` from `internal` to `api`:
![image](https://user-images.githubusercontent.com/10003183/97442604-5298d280-192a-11eb-8875-86d8cb378793.png)
4. Other classes in `com.datastax.oss.driver.internal.core.metadata.token` used `TokenFactory`. Because before the change it was in the same package, no import was needed. After moving `TokenFactory` from `internal` to `api`, this import is needed.

haaawk replied:
Ok makes sense but why isn't it a separate commit which describes what's going on and why?

avelanarius replied:
Separated into a52dc57efcd1ff7e1026ec0578081b8e175c9a1c, with description in the commit message.

haaawk commented on core/src/main/java/com/datastax/oss/driver/internal/core/metadata/token/RandomTokenFactory.java:
```diff
@@ -16,6 +16,7 @@
 package com.datastax.oss.driver.internal.core.metadata.token;
 
 import com.datastax.oss.driver.api.core.metadata.token.Token;
+import com.datastax.oss.driver.api.core.metadata.token.TokenFactory;
```
Why is this needed when you haven't changed this class?

haaawk commented on core/src/main/java/com/datastax/oss/driver/internal/core/metadata/token/Murmur3TokenFactory.java:
```diff
@@ -16,6 +16,7 @@
 package com.datastax.oss.driver.internal.core.metadata.token;
 
 import com.datastax.oss.driver.api.core.metadata.token.Token;
+import com.datastax.oss.driver.api.core.metadata.token.TokenFactory;
```
Why is this needed when you haven't changed this class?

haaawk commented on core/src/main/java/com/datastax/oss/driver/internal/core/metadata/token/Murmur3Token.java:
```diff
@@ -15,49 +15,14 @@
-public class Murmur3Token implements Token {
-
-  private final long value;
+public class Murmur3Token extends TokenLong64 {
```
This extraction should be in a separate commit for the sake of easier review.

avelanarius replied:
Separated into separate commit.

haaawk commented on core/src/main/java/com/datastax/oss/driver/internal/core/metadata/token/TokenFactoryRegistry.java:
```diff
@@ -15,6 +15,8 @@
  */
 package com.datastax.oss.driver.internal.core.metadata.token;
 
+import com.datastax.oss.driver.api.core.metadata.token.TokenFactory;
```
Why is this needed when you haven't changed this class?

haaawk commented on core/src/main/java/com/datastax/oss/driver/internal/core/metadata/token/CDCTokenRange.java:
```diff
@@ -0,0 +1,48 @@
+    }
+
+    List<BigInteger> values = super.split(start, range, RING_END, RING_LENGTH, numberOfSplits);
+    List<Token> tokens = Lists.newArrayListWithExpectedSize(values.size());
+    for (BigInteger value : values) {
+      tokens.add(new CDCToken(value.longValue()));
+    }
+    return tokens;
```
You could replace this with `values.stream().map(v -> new CDCToken(v.longValue())).collect(Collectors.<Token>toList())` or `values.stream().mapToLong(v -> v.longValue()).map(CDCToken::new).collect(Collectors.<Token>toList())`

avelanarius replied:
Fixed in amended commit.

haaawk commented on core/src/main/java/com/datastax/oss/driver/internal/core/metadata/token/DefaultTokenFactoryRegistry.java:
```diff
@@ -15,6 +15,7 @@
  */
 package com.datastax.oss.driver.internal.core.metadata.token;
 
+import com.datastax.oss.driver.api.core.metadata.token.TokenFactory;
```
Why is this needed when you haven't changed this class?

haaawk commented on core/src/main/java/com/datastax/oss/driver/api/core/metadata/token/TokenFactory.java:
```diff
@@ -13,10 +13,8 @@
-package com.datastax.oss.driver.internal.core.metadata.token;
+package com.datastax.oss.driver.api.core.metadata.token;
 
-import com.datastax.oss.driver.api.core.metadata.token.Token;
-import com.datastax.oss.driver.api.core.metadata.token.TokenRange;
 import java.nio.ByteBuffer;
```
Why?

haaawk commented on core/src/main/java/com/datastax/oss/driver/api/core/metadata/TokenMap.java:
```diff
@@ -51,12 +52,19 @@
    */
   @NonNull
-  Token newToken(@NonNull ByteBuffer... partitionKey);
+  Token newToken(TokenFactory partitioner, @NonNull ByteBuffer... partitionKey);
```
Maybe annotate partitioner as Nullable?

avelanarius replied:
Done in amended commits.

haaawk commented on core/src/main/java/com/datastax/oss/driver/internal/core/metadata/DefaultMetadata.java:
```diff
@@ -20,10 +20,10 @@
 import com.datastax.oss.driver.api.core.metadata.Node;
 import com.datastax.oss.driver.api.core.metadata.TokenMap;
 import com.datastax.oss.driver.api.core.metadata.schema.KeyspaceMetadata;
+import com.datastax.oss.driver.api.core.metadata.token.TokenFactory;
```
Please don't do changes like this. They will create merge conflicts later and give us absolutely no value.

haaawk commented on core/src/main/java/com/datastax/oss/driver/internal/core/metadata/FullNodeListRefresh.java:
```diff
@@ -16,9 +16,9 @@
 package com.datastax.oss.driver.internal.core.metadata;
 
 import com.datastax.oss.driver.api.core.metadata.Node;
+import com.datastax.oss.driver.api.core.metadata.token.TokenFactory;
```
Please revert.

haaawk commented on core/src/main/java/com/datastax/oss/driver/internal/core/metadata/InitialNodeListRefresh.java:
```diff
@@ -17,8 +17,8 @@
 
 import com.datastax.oss.driver.api.core.metadata.EndPoint;
 import com.datastax.oss.driver.api.core.metadata.Node;
+import com.datastax.oss.driver.api.core.metadata.token.TokenFactory;
```
Please revert.

haaawk commented:
The change looks ok at the first sight but there's so much noise that it's hard to focus on the important bits.
Please remove all the changes that are not relevant to the addition of CDC partitioner. Namely:
1. Reformating code that hasn't changed
2. Reordering imports
3. Changing imports even though the class hasn't been touched.
4. Anything else that's not strictly needed for CDC partitioner support.

Also please do Murmur3Token -> Long64Token refactoring as a separate commit.

avelanarius committed:
Add TokenLong64 class

Introduces new TokenLong64 class, which is an abstraction of all tokens
that can be represented as 64-bit integers. Change ShardingInfo to take
advantage of this new class.

avelanarius committed:
Add CDC token implementation

Adds CDCToken, CDCTokenFactory and CDCTokenRange classes, which
implement a Scylla CDC partitioner, as defined here:

https://github.com/scylladb/scylla/blob/master/cdc/cdc_partitioner.cc

avelanarius committed:
Add getPartitioner() to Request, derived classes

Adds new getPartitioner() method to allow to specify a custom
partitioner to be used on a per-Request basis.

Modify Conversions to set CDCPartitioner for CDC log prepared
statements.

avelanarius commented:
@haaawk Split commit into multiple smaller commits. 

Regarding points 1-4: All those imports are related mostly to moving `TokenFactory` from `internal` to `api`. When I was splitting the commits, I went practically over the every line once again and verified that this is the case.

Also, as to reformatting and reordering imports - it's not my mistake or some IDE left-over: the style checker in build system **very strictly** enforces this - including reordering imports and deleting old (unused) ones.

haaawk reviewed (changes requested):
Thanks @avelanarius it's much easier to review now.
The change looks good but I would like to suggest some improvements:
1. Don't move TokenFactory class to api - this is a very intrusive change which I expect to cause a lot of pain in the future while we will be merging with upstream. Please instead define a new Partition interface that's passed around and has the hash method. Internally implemented using TokenFactory but without exposing it.
2.  Please extract partitioner method from Conversions.java into helper class and use this helper class inside Conversions.java. Just to minimize the risk of merge conflicts.

Otherwise the change LGTM

avelanarius commented:
@haaawk Regarding the first point, is this a good way to do it?:

`TokenFactory` is already an interface, so change it to abstract class which will `implements` new `Partition` interface. This `TokenFactory` change will necessitate changing `implements` to `extends` in all token factories classes (a few files). `Partition` interface will consist of all methods of `TokenFactory`. 

To limit the lines of code/file count changed, the only place that will use `Partition` will be `getPartition()`. In all other places `getPartition()` will immediately be cast to `TokenFactory` (other way would require changing `TokenFactory` to `Partition` in large number of places).

avelanarius committed:
Add Partitioner interface

Adds new Partitioner interface which is public-facing (in api package),
compared to TokenFactory which was an internal class. Partitioner
interface constists of single hash method.

Change TokenFactory to extend the new Partitioner interface.

avelanarius committed:
Add PartitionerFactory class

Add helper class PartitionerFactory, which returns correct Partitioner
for a given (prepared statement) query (as specified by variable 
definitions).

avelanarius committed:
Use custom partitioner in TokenMap and handler

Change TokenMap, CqlRequestHandler to be able to use custom partitioner,
as specified in getPartitioner() method of Request.

avelanarius commented:
@haaawk Pushed amended commits - fixed those two points you mentioned.

haaawk commented on core/src/main/java/com/datastax/oss/driver/internal/core/metadata/token/TokenFactory.java:
```diff
@@ -15,26 +15,23 @@
-import java.nio.ByteBuffer;
 
 /** Manages token instances for a partitioner implementation. */
-public interface TokenFactory {
```
I don't understand why this change is needed. Why can't TokenFactory by an interface that extends Partitioner interface?

haaawk replied:
Even better would be for TokenFactory.getPartitionerName to be converted into TokenFactory.getPartitioner that returns Partitioner. I'm not sure whether it would fit well with the code though.

haaawk commented on core/src/main/java/com/datastax/oss/driver/internal/core/metadata/token/TokenFactory.java:
```diff
@@ -15,26 +15,23 @@
+import com.datastax.oss.driver.api.core.metadata.token.Partitioner;
 import com.datastax.oss.driver.api.core.metadata.token.Token;
 import com.datastax.oss.driver.api.core.metadata.token.TokenRange;
-import java.nio.ByteBuffer;
```
Please don't do unrelated cleanups.

avelanarius replied:
Again, this is caused by automatic style checker:
```
[ERROR] Found 1 non-complying files, failing build
[ERROR] To fix formatting errors, run "mvn com.coveo:fmt-maven-plugin:format"
```
(after adding back this import)

haaawk replied:
ok.

haaawk commented on core/src/main/java/com/datastax/oss/driver/internal/core/metadata/token/DefaultTokenMap.java:
```diff
@@ -171,9 +176,14 @@ public TokenRange newTokenRange(@NonNull Token start, @NonNull Token end) {
-    return (keyspaceMap == null) ? Collections.emptySet() : keyspaceMap.getReplicas(partitionKey);
+    return (keyspaceMap == null)
+        ? Collections.emptySet()
+        : keyspaceMap.getReplicas((TokenFactory) partitioner, partitionKey);
```
Please add a comment explaining why it is safe to cast to TokenFactory here.

haaawk commented on core/src/main/java/com/datastax/oss/driver/internal/core/metadata/token/KeyspaceTokenMap.java:
```diff
@@ -96,8 +96,11 @@ private KeyspaceTokenMap(
 
-  Set<Node> getReplicas(ByteBuffer partitionKey) {
-    return getReplicas(tokenFactory.hash(partitionKey));
+  Set<Node> getReplicas(TokenFactory partitioner, ByteBuffer partitionKey) {
```
Why not to take Partitioner instead of TokenFactory here and remove casting from DefaultTokenMap.getReplicas?

haaawk commented:
LGTM in general but I don't understand why TokenFactory can't be an interface anymore.
Also, I think it might be possible to remove casting around getReplicas.

avelanarius commented:
Changed `TokenFactory` back to `interface` (previously was under misunderstanding that interfaces cannot extend other interfaces...). Removed casting around `getReplicas`.

haaawk commented on core/src/main/java/com/datastax/oss/driver/api/core/metadata/TokenMap.java:
```diff
@@ -99,15 +108,39 @@
   /** The replicas for a given partition key in the given keyspace. */
   @NonNull
-  Set<Node> getReplicas(@NonNull CqlIdentifier keyspace, @NonNull ByteBuffer partitionKey);
+  Set<Node> getReplicas(
+      @NonNull CqlIdentifier keyspace,
+      @Nullable Partitioner partitioner,
+      @NonNull ByteBuffer partitionKey);
+
+  /**
+   * Shortcut for {@link #getReplicas(CqlIdentifier, Partitioner, ByteBuffer) getReplicas(keyspace,
+   * null, partitionKey)}.
+   */
+  @NonNull
+  default Set<Node> getReplicas(@NonNull CqlIdentifier keyspace, @NonNull ByteBuffer partitionKey) {
+    return getReplicas(keyspace, null, partitionKey);
+  }
+
+  /**
+   * Shortcut for {@link #getReplicas(CqlIdentifier, Partitioner, ByteBuffer)
+   * getReplicas(CqlIdentifier.fromCql(keyspaceName), partitioner, partitionKey)}.
+   */
+  @NonNull
+  default Set<Node> getReplicas(
+      @NonNull String keyspaceName,
+      @Nullable Partitioner partitioner,
+      @NonNull ByteBuffer partitionKey) {
+    return getReplicas(CqlIdentifier.fromCql(keyspaceName), partitioner, partitionKey);
+  }
```
since we've got those shortcuts, are really all those changes in the tests necessary?

avelanarius replied:
Yes, I think so. I think that those tests directly call `getReplicas(keyspace, partitioner, partitionKey)` with `partitioner = null`, so mocks had to be updated. Without those shortcuts even more places would have to be fixed.

haaawk commented on core/src/test/java/com/datastax/oss/driver/internal/core/loadbalancing/BasicLoadBalancingPolicyQueryPlanTest.java:
```diff
@@ -90,7 +91,9 @@ public void should_use_round_robin_when_no_request() {
-    then(tokenMap).should(never()).getReplicas(any(CqlIdentifier.class), any(ByteBuffer.class));
+    then(tokenMap)
+        .should(never())
+        .getReplicas(any(CqlIdentifier.class), isNull(), any(ByteBuffer.class));
```
Shouldn't isNull() be any()?

haaawk replied:
And in other should(never()) places.
